docs/extensions/storage-transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,293 @@
+`key` is a sequence of characters and a `value` is a sequence of bytes. A
+key-value pair is called `entry` in the following part.
+
+This sharding transformer only adapts entries where the key starts with
+`data/root`, as they indicate data keys for array chunks, see
+:ref:`storage-keys`. All other entries are simply passed on.
+
+Entries starting with ``data/root`` are grouped by their common shard, assuming
+storage keys from a regular chunk grid which may use a custom configured
+``chunk separator``: For all entries that are part of the same shard the key is
+changed to the shard-key and the values are combined in the
+`Binary shard format`_ as described below. The new shard-key is the chunk key
+divided by ``chunks_per_shard`` and floored per dimension. For example for
+``chunks_per_shard=[32, 2]``, the chunk grid position ``[96, 18]`` (e.g. key
+"data/root/foo/baz/c96/18") is transformed to the shard grid position
+``[3, 9]`` and reassigned to the respective new key, honoring the original chunk
+separator (e.g. "data/root/foo/baz/c3/9"). Chunk grid positions ``[96, 19]``,
+``[97, 18]``, …, up to ``[127, 19]`` will also have the same shard grid position
+``[3, 9]``.
+
+
+Binary shard format
+-------------------
+
+The only binary format is the ``indexed`` format, as specified by the ``type``
+configuration key. Other binary formats might be added in future versions.
+
+In the indexed binary format, chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them. The
+index is placed at the end of the file and has a size of 16 bytes multiplied by
+the number of chunks in a shard, for example ``16 bytes * 64 = 1024 bytes`` for
+``chunks_per_shard=[32, 2]``. The index holds an `offset, nbytes` pair of
+little-endian uint64 per chunk, the chunks-order in the index is row-major (C)
+order, for example for ``chunks_per_shard=[2, 2]`` an index would look like:
+
+.. code-block::
+
+    | chunk (0, 0)    | chunk (0, 1)    | chunk (1, 0)    | chunk (1, 1)    |
+    | offset | nbytes | offset | nbytes | offset | nbytes | offset | nbytes |
+    | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 |
+
+
+Empty chunks are denoted by setting both offset and nbytes to ``2^64 - 1``. The
+index always has the full shape of all possible chunks per shard, even if they
+are outside of the array size.
+
+The actual order of the chunk content is not fixed and may be chosen by the
+implementation. All possible write orders are valid according to this
+specification and therefore can be read by any other implementation. When
+writing partial chunks into an existing shard, no specific order of the existing
+chunks may be expected. Some writing strategies might be
+
+* **Fixed order**: Specify a fixed order (e.g. row-, column-major, or Morton
+  order). When replacing existing chunks larger or equal-sized chunks may be
```

Comment:
cc @glemson who is interested in space-filling curves in Zarr

Reply:
@glemson - I hope you and the SciServer team will be able to provide some feedback on this plan before it is finalized.

Reply:
We have people working on applying zarr to data form the JHU turbulence database and comparing it to our existing solutions using space filling curves. we want to include their conclusions in our commenting. Sorry this does not go faster.

Reply:
@rabernat @glemson Using a space-filling curve here, I think, makes this analogous to our existing solution for the JHU Turbulence data. So I think doing it this way would make testing/transitioning our current data/processes to zarr simpler.

Reply:
@glemson @ryanhausen Thanks for the input! In principle the current spec would fit to your use case, the only thing is to ensure that the writer uses the wished order for your case, so this is depending on the concrete implementation. Out of interest, what implementation are you currently considering to use for zarr?

Reply:
@jstriebel Of course! This is a great project. I agree that it depends on the concrete implementation. Right now we're prototyping and doing some performance testing so we haven't settled on an implementation in particular. I think we're willing to explore changing how we do things to better fit zarr in exchange for the simplicity and comparable (or improved) performance of using zarr.